x/bank/keeper/genesis.go

```diff
@@ -10,8 +10,14 @@ import (
 
 // InitGenesis initializes the bank module's state from a given genesis state.
 func (k BaseKeeper) InitGenesis(ctx sdk.Context, genState *types.GenesisState) {
+	genState.MigrateSendEnabled()
```

Comment:
Why is this being called here? Migrations should happen explicitly in upgrade handlers.

Reply:
This allows continued use of a genesis file that has `SendEnabled` entries defined in the bank params.

I've updated genesis file creation to have any new genesis files created using the new field, but I'm unaware of any code location meant for migrating a genesis file. By having this line here, though, there's no need to migrate the file.

All `MigrateSendEnabled` does is move things around in the GenesisState object (from `GenesisState .Params.SendEnabled` to the new `GenesisState.SendEnabled` field). It doesn't update any files.

Reply:
I see. I like how you thought of this, but this is not how it's supposed to work.

`InitGenesis` is supposed to work as if it's working on the latest valid state. This is why we have "migrations". You can see many examples of this in some of the modules. In fact, `x/bank` already has existing migrations. See `RegisterServices`. You'll find various `RegisterMigration` calls there. We need to do the same for this. So we have to remove `genState.MigrateSendEnabled()` :)

Reply:
I've already added/registered a new migration for the bank module to migrate the `Params.SendEnabled` entries into the bank module's state store: 

[x/bank/keeper/migrations.go](https://github.com/cosmos/cosmos-sdk/pull/11977/files#diff-c58b556b8ace809b59cce155271115f96a755f2d04d3ac9368c5c58e5791bc1e):
```go
// Migrate3to4 migrates x/bank storage from version 3 to 4.
func (m Migrator) Migrate3to4(ctx sdk.Context) error {
	oldParams := m.keeper.GetParams(ctx)
	m.keeper.SetAllSendEnabled(ctx, oldParams.GetSendEnabled())
	m.keeper.SetParams(ctx, banktypes.NewParams(oldParams.DefaultSendEnabled))
	return nil
}
```
and [x/bank/module.go](https://github.com/cosmos/cosmos-sdk/pull/11977/files#diff-5d9b18806fe02e80ffd38cea841e692ce20ef3d8941b135477658bad46ed9937):
```go
	if err := cfg.RegisterMigration(types.ModuleName, 3, m.Migrate3to4); err != nil {
		panic(fmt.Sprintf("failed to migrate x/bank from version 3 to 4: %v", err))
	}
```

I don't see the harm in having the `genState.MigrateSendEnabled()` call in here. All it does is make the current genesis file `Params.SendEnabled` entries work like the new `SendEnabled` entries. All that function does is move things around in that `genState` struct. It puts the `SendEnabled` entries where they're expected to be so that later in `InitGenesis`, we only need to look in one place for those entries. It allows for backwards compatibility with the old structure (defining them in Params), even in the genesis files.

Reply:
Ahh I see! Thank you.

Re `InitGenesis`, we don't want any migrations, that's the whole point of explicit migrations. We want app devs to explicitly execute migrations and not have it happen magically via `InitGenesis`. In fact, `InitGenesis` would only in this case would only be called for a new chain, in which case there is nothing to migrate anyway.

Please remove this so we can wrap up the PR 🙏